aten/src/ATen/native/cuda/Blas.cpp

```diff
@@ -999,14 +1010,22 @@ ScalingType get_scaling_type(
     int64_t dim_m,
     int64_t dim_k,
     int64_t dim_n) {
-  // Check for BlockWise scaling (FP8_E8M0 types)
-  if (scale_a.scalar_type() == scale_b.scalar_type() &&
-      scale_a.scalar_type() == at::kFloat8_e8m0fnu) {
-    constexpr int64_t BLOCK_SIZE_K = 32;
+  // Check for BlockWise scaling (FP8_E8M0 and FP8_E4M3 types)
+  if ((scale_a.scalar_type() == scale_b.scalar_type()) &&
+      ((scale_a.scalar_type() == at::kFloat8_e8m0fnu) || (scale_a.scalar_type() == at::kFloat8_e4m3fn))) {
+    const bool is_nvfp4 = scale_a.scalar_type() == at::kFloat8_e4m3fn;
+
+    // cuBLAS's mxfp8 gemm: block_size is 1 scale per 32 elements
+    // cuBLAS's nvfp4 gemm: block_size is 1 scale per 16 unpacked elements.
+    const auto BLOCK_SIZE_K = is_nvfp4 ? 16 : 32;
+
     constexpr int64_t BLOCK_SIZE_MN = 128;
 
+    // adjust for fp4x2 packing if necessary
+    const auto dim_k_unpacked = is_nvfp4 ? dim_k * 2 : dim_k;
+
     auto ceil_div = [](auto a, auto b) { return (a + b - 1) / b; };
-    auto num_k_blocks = ceil_div(dim_k, BLOCK_SIZE_K);
+    auto num_k_blocks = ceil_div(dim_k_unpacked, BLOCK_SIZE_K);
```

Comment:
nit: maybe just pass in the unpacked dim_k

Reply:
sorry, could you clarify the suggestion?

Reply:
Ohh I meant just handle the unpacking in the calling func

Reply:
hmm, I took a look and it seems a bit awkward to do the unpacking in `_scaled_mm_out_cuda`, because that function has logic which uses the packed number...is that what you meant? Maybe I'm still misunderstanding.